File: LeetCode-August-Challenge/BinaryTreesWithFactors.java
import java.util.*;
class Solution {
    public int numFactoredBinaryTrees(int[] A) {
        long res = 0L, mod = (long)1e9 + 7;
        Arrays.sort(A);
        HashMap<Integer, Long> dp = new HashMap<>();
        for (int i = 0; i < A.length; ++i) {
            dp.put(A[i], 1L);
            for (int j = 0; j < i; ++j)
                if (A[i] % A[j] == 0)
                    dp.put(A[i], (dp.get(A[i]) + dp.get(A[j]) * dp.getOrDefault(A[i] / A[j], 0L)) % mod);
            res = (res + dp.get(A[i])) % mod;
        }
        return (int) res;
    }
}
public class BinaryTreesWithFactors {
    public static void main(String[] args) {
        int[] arr = {2,4};
        Solution s = new Solution();
        System.out.println(s.numFactoredBinaryTrees(arr));
    }
}

File: Year-2022/LeetCode-August-Challenge/BinaryTreesWithFactors.cpp
#include<bits/stdc++.h>
using namespace std;
static int MOD = 1e9 + 7;
class Solution {
public:
    int numFactoredBinaryTrees(vector<int>& arr) {
        map<int, long long int> mp; int n = arr.size();
        sort(arr.begin(), arr.end());
        for(int i = 0; i < n; ++i){
            mp.insert({arr[i], 1});
        }
        for(int i = 1; i < n; ++i){
            auto it3 = mp.find(arr[i]);
            long long int count = 0;
            for(int j = 0; j < i; ++j){
                if(arr[i] % arr[j] == 0){
                    auto it = mp.find(arr[i]/arr[j]);
                    auto it2 = mp.find(arr[j]);
                    if(it != mp.end()){
                        count += (it->second * it2->second);
                    }
                }
            }
            it3->second += count;
        }
        long long int res = 0;
        for(auto parts : mp){
            res += parts.second;
        }
        return res % MOD;
    }
};
int main(){
    vector<int> arr = {2,4};
    Solution s;
    cout<<s.numFactoredBinaryTrees(arr)<<endl;
    return 0;
}